specs/DesignDiscussions/DotNetTemplate.md

```diff
@@ -0,0 +1,338 @@
+```json
+{
+	"$schema": "http://adaptivecards.io/schemas/adaptive-card.json",
+	"type": "AdaptiveCard",
+	"version": "1.0",
+	"body": [
+		{
+			"type": "TextBlock",
+			"text": "**wrap: true** Lorem ipsum dolor sit amet",
+			"wrap": true
+		}
+	]
+}
+```
+The above json will be parsed into a parse tree by the parser.  
+ANTLR also provides a walker that takes a visitor as an argument. As the walker walks the tree, visitor's visit methods are called. Visit methods are defined per new grammar by default.
+When a visitor method is called, it will be called with a context. The context will be a data structure representing the current visiting node.
+
+```
+pair
+   : StringDeclOpen STRING CLOSE COLON value    # jsonPair
+   | StringDeclOpen TEMPLKEYWRD CLOSE COLON value # templateData
+   ;
+```
+
+From this grammar snippet, the parser generation tool will create a pairVisit method to a base visitor that we have to implement. By overriding this method, when walker visit the pair node, it calls the visit method of the node. We can do our own work such as translation in the visit method.
+
+**Figure 2**
+![](antlr4_parse_tree_1.png)
+
+**Parse Tree with Templated String**
+```json
+{
+	"$schema": "http://adaptivecards.io/schemas/adaptive-card.json",
+	"type": "AdaptiveCard",
+	"version": "1.0",
+	"body": [
+		{
+			"type": "TextBlock",
+			"text": "**wrap: true** {content}",
+			"wrap": true
+		}
+	]
+}
+```
+
+Notice that templateString node has been added, likewise other template keywords such as $data, $when, $root, $index can be parsed by the generated parser and lexer  
+Parser and lexer can be used on other platforms since they are generated by ANTLR tool and can be targeted to different languages. Visitor and Return Objects are needed to be implemented on other platforms.  
+  
+**Figure 3**
+![](antlr4_parse_tree_2.png)
+
+## Adaptive Expression Language
+This was initially intended to provide us functions support. AEL already has many built-in functions such as toLower, toString, equals, add, div, and [etc](https://github.com/microsoft/BotBuilder-Samples/blob/master/experimental/common-expression-language/prebuilt-functions.md)  
+But the AEL can do more. It can evaluate our templated string as well.
+
+```dotnet
+private ExpressionEngine _expressionEngine = new ExpressionEngine();
+/// some lines of code later
+var result = _expressionEngine.Parse(unboundString).TryEvaluate(data);
+```
+
+**Signature of TryEvaluate**
+```dotnet
+/// <summary>
+/// Evaluate the expression.
+/// </summary>
+/// <param name="state">
+/// Global state to evaluate accessor expressions against.  Can be <see cref="IDictionary{String}{Object}"/>, <see cref="IDictionary"/> otherwise reflection is used to access property and then indexer.
+/// </param>
+/// <returns>Computed value and an error string.  If the string is non-null, then there was an evaluation error.</returns>
```

Comment:
>f the string is non-nul [](start = 51, length = 23)

Discussion in room: These comments seem not entirely up to date.

Reply:
Add a link to wherever this came from.

---
In reply to: [382839131](https://github.com/microsoft/AdaptiveCards/pull/3798#discussion_r382839131) [](ancestors = 382839131)

Reply:
Nevermind, the comment is fine, but adding a link would still help.

---
In reply to: [382839335](https://github.com/microsoft/AdaptiveCards/pull/3798#discussion_r382839335) [](ancestors = 382839335,382839131)